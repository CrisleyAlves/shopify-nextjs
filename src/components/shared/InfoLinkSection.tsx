import Link from "next/link";

import type { MenuType } from "@/lib/shopify/types";

import Icon from "./Icon";

type IconPath = "/about" | "/privacy" | "/contact-us" | "/guarantee";

const Icons: Record<IconPath, string> = {
  "/about": "/icons/account.svg",
  "/privacy": "/icons/eye.svg",
  "/contact-us": "/icons/email-symbol.svg",
  "/guarantee": "/icons/security.svg",
};

/**
 * @todo make it dynamic -- i.e missing guarantee page in Shopify
 */
export default function InfoLinkSection({
  sections,
}: {
  sections: MenuType[];
}): React.JSX.Element {
  return (
    <section className="text-black font-extralight pt-10 pb-10 border-t-2 bg-gray-50">
      <ul className="grid grid-cols-2 gap-y-5 md:gap-y-0 md:grid-cols-4 md:container">
        {sections.map((section) => {
          const icon = Icons[section.path as IconPath];

          return (
            <li
              key={section.path}
              className="flex flex-col items-center justify-center"
            >
              <Icon path={icon} altText={section.title} />
              <h3 className="text-sm mt-2 capitalize">{section.title}</h3>
              <Link
                href={section.path}
                className="text-sm underline text-gray-500"
              >
                Learn More
              </Link>
            </li>
          );
        })}
      </ul>
    </section>
  );
}
