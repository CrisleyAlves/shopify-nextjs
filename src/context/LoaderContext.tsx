"use client";
import { createContext, useContext, useEffect, useState } from "react";
import { usePathname, useRouter, useSearchParams } from "next/navigation";

interface LoaderProviderType {
  showLoader: boolean;
  setShowLoader: (value: boolean) => void;
  pushWithLoader: (url: string) => void;
}

const LoaderContext = createContext<LoaderProviderType | undefined>(undefined);

export function LoaderProvider({ children }: { children: React.ReactNode }) {
  const [showLoader, setShowLoader] = useState(false);

  const router = useRouter();
  const pathanme = usePathname();
  const searchParams = useSearchParams();

  useEffect(() => {
    setTimeout(() => {
      setShowLoader(false);
    }, 500);
  }, [pathanme, searchParams]);

  const pushWithLoader = (url: string) => {
    setShowLoader(true);
    router.push(url);
  };

  return (
    <LoaderContext.Provider
      value={{
        showLoader,
        setShowLoader,
        pushWithLoader,
      }}
    >
      {children}
    </LoaderContext.Provider>
  );
}

export function useLoader() {
  const context = useContext(LoaderContext);
  if (!context) {
    throw new Error("useLoader must be used within a LoaderProvider");
  }
  return context;
}
