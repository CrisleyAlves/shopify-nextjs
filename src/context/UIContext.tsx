"use client";
import { createContext, useContext, useState } from "react";

interface UIProviderType {
  showStickyNav: boolean;
  setShowStickyNav: (value: boolean) => void;
  showSearchForm: boolean;
  setShowSearchForm: (value: boolean) => void;
  showCart: boolean;
  setShowCart: (value: boolean) => void;
  showSidenav: boolean;
  setShowSidenav: (value: boolean) => void;
  showLoader: boolean;
  setShowLoader: (value: boolean) => void;
}

const UIContext = createContext<UIProviderType | undefined>(undefined);

export function UIProvider({ children }: { children: React.ReactNode }) {
  const [showStickyNav, setShowStickyNav] = useState(false);
  const [showSearchForm, setShowSearchForm] = useState(false);
  const [showCart, setShowCart] = useState(false);
  const [showSidenav, setShowSidenav] = useState(false);
  const [showLoader, setShowLoader] = useState(false);

  return (
    <UIContext.Provider
      value={{
        showStickyNav,
        setShowStickyNav,
        showSearchForm,
        setShowSearchForm,
        showCart,
        setShowCart,
        showSidenav,
        setShowSidenav,
        showLoader,
        setShowLoader,
      }}
    >
      {children}
    </UIContext.Provider>
  );
}

export function useUI() {
  const context = useContext(UIContext);
  if (!context) {
    throw new Error("useUI must be used within a UIProvider");
  }
  return context;
}
