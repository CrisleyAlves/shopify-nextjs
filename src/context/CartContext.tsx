"use client";
import {
  createContext,
  ReactElement,
  useContext,
  useMemo,
  useState,
} from "react";

import type { Cart, CartItem } from "@/lib/shopify/cart/types";
import type { ProductVariant } from "@/lib/shopify/product/types";

import { MESSAGES, NOTIFICATION_TYPES } from "@/constants";
import { addItemToCart, updateItemQuantity } from "@/services/cart-service";

import { useUI } from "./UIContext";

type CartContextType = {
  cart: Cart | undefined;
  updateShopifyCart: (payload: any) => void;
  addToCart: (selectedVariant: ProductVariant) => void;
  increaseItemQuantity: (item: CartItem) => void;
  decreaseItemQuantity: (item: CartItem) => void;
  totalQuantity: number;
  isEmpty: boolean;
};

const CartContext = createContext<CartContextType | undefined>(undefined);

export function CartProvider({
  children,
  shopifyCart,
}: {
  children: React.ReactNode;
  shopifyCart: Cart | undefined;
}): ReactElement {
  const [cart, setCart] = useState<Cart | undefined>(shopifyCart);
  const { setShowLoader, handleNotification } = useUI();

  const updateShopifyCart = (payload: Cart) => setCart(payload);

  const addToCart = async (selectedVariant: ProductVariant) => {
    try {
      setShowLoader(true);
      const shopifyCart = await addItemToCart(selectedVariant.id);

      if (shopifyCart instanceof Error) {
        throw shopifyCart;
      }

      updateShopifyCart(shopifyCart);
      handleNotification({
        type: NOTIFICATION_TYPES.SUCCESS,
        message: MESSAGES.SUCCESS.ADD_TO_CART,
        visible: true,
      });
    } catch (error) {
      handleNotification({
        type: NOTIFICATION_TYPES.ERROR,
        message: MESSAGES.ERROR.ADD_TO_CART,
        visible: true,
      });
      console.error("Error adding item to cart");
      console.error(error);
    } finally {
      setShowLoader(false);
    }
  };

  const increaseItemQuantity = async (item: CartItem) => {
    try {
      setShowLoader(true);
      const shopifyCart = await updateItemQuantity({
        merchandiseId: item.merchandise.id,
        quantity: item.quantity + 1,
      });

      if (shopifyCart instanceof Error) {
        throw shopifyCart;
      }

      updateShopifyCart(shopifyCart);

      handleNotification({
        type: NOTIFICATION_TYPES.SUCCESS,
        message: MESSAGES.SUCCESS.INCREASE_QUANTITY,
        visible: true,
      });
    } catch (error) {
      handleNotification({
        type: NOTIFICATION_TYPES.ERROR,
        message: MESSAGES.ERROR.INCREASE_QUANTITY,
        visible: true,
      });
      console.error("Error increasing item quantity");
      console.error(error);
    } finally {
      setShowLoader(false);
    }
  };

  const decreaseItemQuantity = async (item: CartItem) => {
    try {
      setShowLoader(true);
      const shopifyCart = await updateItemQuantity({
        merchandiseId: item.merchandise.id,
        quantity: item.quantity - 1,
      });

      if (shopifyCart instanceof Error) {
        throw shopifyCart;
      }

      updateShopifyCart(shopifyCart);

      const decreaseMessage =
        item.quantity === 1
          ? MESSAGES.SUCCESS.ITEM_REMOVED_FROM_CART
          : MESSAGES.SUCCESS.DECREASE_QUANTITY;
      handleNotification({
        type: NOTIFICATION_TYPES.SUCCESS,
        message: decreaseMessage,
        visible: true,
      });
    } catch (error) {
      handleNotification({
        type: NOTIFICATION_TYPES.ERROR,
        message: MESSAGES.ERROR.DECREASE_QUANTITY,
        visible: true,
      });
      console.error("Error decreasing item quantity");
      console.error(error);
    } finally {
      setShowLoader(false);
    }
  };

  const value = useMemo(
    () => ({
      cart,
      updateShopifyCart,
      increaseItemQuantity,
      decreaseItemQuantity,
      addToCart,
      isEmpty: (cart?.lines?.length || 0) === 0,
      totalQuantity: cart?.totalQuantity || 0,
    }),
    [cart]
  );

  return <CartContext.Provider value={value}>{children}</CartContext.Provider>;
}

export function useCart() {
  const context = useContext(CartContext);
  if (!context) throw new Error("useCart must be used within a CartProvider");
  return context;
}
