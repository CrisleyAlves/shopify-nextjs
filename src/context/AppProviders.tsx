import type { Cart } from "@/lib/shopify/cart/types";

import { CartProvider } from "./CartContext";
import { UIProvider } from "./UIContext";

export function AppProviders({
  children,
  shopifyCart,
}: {
  children: React.ReactNode;
  shopifyCart: Cart | undefined;
}) {
  return (
    <UIProvider>
      <CartProvider shopifyCart={shopifyCart}>{children}</CartProvider>
    </UIProvider>
  );
}
