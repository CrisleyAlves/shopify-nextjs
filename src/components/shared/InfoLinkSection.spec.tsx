import { render, screen } from "@testing-library/react";
import { LoaderProvider } from "@/context/LoaderContext";

import InfoLinkSection from "./InfoLinkSection";

const mockedSections = [
  {
    path: "/about",
    title: "About",
  },
  {
    path: "/privacy",
    title: "Privacy",
  },
  {
    path: "/contact-us",
    title: "Contact Us",
  },
  {
    path: "/guarantee",
    title: "Guarantee",
  },
];

/**
 * @todo: move to a global config
 */
jest.mock("next/navigation", () => ({
  useRouter: () => ({
    push: jest.fn(),
    replace: jest.fn(),
    prefetch: jest.fn(),
    pathname: "/",
    query: {},
    asPath: "/",
  }),
  useSearchParams: () => ({
    get: jest.fn(),
    has: jest.fn(),
    entries: jest.fn(),
    keys: jest.fn(),
    values: jest.fn(),
    toString: jest.fn(),
  }),
  usePathname: () => "/",
}));

describe("InfoLinkSection", () => {
  it("should render section correcly", () => {
    const { container } = render(
      <LoaderProvider>
        <InfoLinkSection sections={mockedSections} />
      </LoaderProvider>
    );
    expect(screen.getByText("About")).toBeDefined();
    expect(screen.getByText("Privacy")).toBeDefined();
    expect(screen.getByText("Contact Us")).toBeDefined();
    expect(screen.getByText("Guarantee")).toBeDefined();

    expect(container).toMatchSnapshot();
  });
});
