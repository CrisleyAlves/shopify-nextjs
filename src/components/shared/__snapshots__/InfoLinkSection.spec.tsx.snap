// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`InfoLinkSection should render section correcly 1`] = `
<div>
  <section
    class="text-black font-extralight pt-10 pb-10 border-t-2 bg-gray-50"
  >
    <ul
      class="grid grid-cols-2 gap-y-5 md:gap-y-0 md:grid-cols-4 md:container"
    >
      <li
        class="flex flex-col items-center justify-center"
      >
        <img
          alt="About"
          class="w-5 md:w-6"
          data-nimg="1"
          decoding="async"
          height="0"
          loading="lazy"
          src="/icons/account.svg"
          style="color: transparent;"
          width="0"
        />
        <h3
          class="text-sm mt-2 capitalize"
        >
          About
        </h3>
        <button
          class="text-sm underline text-gray-500"
        >
          Learn More
        </button>
      </li>
      <li
        class="flex flex-col items-center justify-center"
      >
        <img
          alt="Privacy"
          class="w-5 md:w-6"
          data-nimg="1"
          decoding="async"
          height="0"
          loading="lazy"
          src="/icons/eye.svg"
          style="color: transparent;"
          width="0"
        />
        <h3
          class="text-sm mt-2 capitalize"
        >
          Privacy
        </h3>
        <button
          class="text-sm underline text-gray-500"
        >
          Learn More
        </button>
      </li>
      <li
        class="flex flex-col items-center justify-center"
      >
        <img
          alt="Contact Us"
          class="w-5 md:w-6"
          data-nimg="1"
          decoding="async"
          height="0"
          loading="lazy"
          src="/icons/email-symbol.svg"
          style="color: transparent;"
          width="0"
        />
        <h3
          class="text-sm mt-2 capitalize"
        >
          Contact Us
        </h3>
        <button
          class="text-sm underline text-gray-500"
        >
          Learn More
        </button>
      </li>
      <li
        class="flex flex-col items-center justify-center"
      >
        <img
          alt="Guarantee"
          class="w-5 md:w-6"
          data-nimg="1"
          decoding="async"
          height="0"
          loading="lazy"
          src="/icons/security.svg"
          style="color: transparent;"
          width="0"
        />
        <h3
          class="text-sm mt-2 capitalize"
        >
          Guarantee
        </h3>
        <button
          class="text-sm underline text-gray-500"
        >
          Learn More
        </button>
      </li>
    </ul>
  </section>
</div>
`;
